Region inference member constraints (#344)
rust-lang/rustc-guide: src/borrow_check/region_inference/constraint_propagation.md

  - For each constraint C:
    - Update `Values` as needed to satisfy the constraint

[riv]: ../region-inference.html#region-variables

As a simple example, if we have a liveness constraint `R live at E`,
then we can apply `Values(R) = Values(R) union {E}` to make the
constraint be satisfied. Similarly, if we have an outlives constraints
`R1: R2`, we can apply `Values(R1) = Values(R1) union Values(R2)`.
(Member constraints are more complex and we discuss them below.)

In practice, however, we are a bit more clever. Instead of applying
the constraints in a loop, we can analyze the constraints and figure
out the correct order to apply them, so that we only have to apply
each constraint once in order to find the final result.

Similarly, in the implementation, the `Values` set is stored in the
`scc_values` field, but they are indexed not by a *region* but by a
*strongly connected component* (SCC). SCCs are an optimization that
avoids a lot of redundant storage and computation.  They are explained
in the section on outlives constraints.

## Liveness constraints

A **liveness constraint** arises when some variable whose type

> includes a region R is live at some point P. This simply means that

Author: added to glossary

by invoking `constraint_sccs.scc(r)`.

Working in terms of SCCs allows us to be more efficient: if we have a
set of regions `'a...'d` that are part of a single SCC, we don't have
to compute/store their values separarely. We can just store one value
**for the SCC**, since they must all be equal.

If you look over the region inference code, you will see that a number
of fields are defined in terms of SCCs. For example, the
[`scc_values`] field stores the values of each SCC. To get the value
of a specific region `'a` then, we first figure out the SCC that the
region is a part of, and then find the value of that SCC.

[`constraint_sccs`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/borrow_check/nll/region_infer/struct.RegionInferenceContext.html#structfield.constraint_sccs
[`scc_values`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/borrow_check/nll/region_infer/struct.RegionInferenceContext.html#structfield.scc_values

When we compute SCCs, we not only figure out which regions are a
member of each SCC, we also figure out the edges between them. So for example
consider this set of outlives constraints:

```
'a: 'b
'b: 'a

'a: 'c

'c: 'd
'd: 'c
```

> Here we have two SCCs: S0 contains `'a` and `'b`, and S1 contains `'c`

Reviewer: ```suggestion
Here we have two SCCs: `S0` contains `'a` and `'b`, and S1 contains `'c`
```

and `'d`.  But these SCCs are not independent: because `'a: 'c`, that
means that `S0: S1` as well. That is -- the value of `S0` must be a
superset of the value of `S1`. One crucial thing is that this graph of
SCCs is always a DAG -- that is, it never has cycles. This is because
all the cycles have been removed to form the SCCs themselves.

### Applying liveness constraints to SCCs

The liveness constraints that come in from the type-checker are
expressed in terms of regions -- that is, we have a map like
`Liveness: R -> {E}`.  But we want our final result to be expressed
in terms of SCCs -- we can integrate these liveness constraints very
easily just by taking the union:

```
for each region R:

>   let S by the SCC that contains R

Reviewer: ```suggestion
  let S be the SCC that contains R
```

  Values(S) = Values(S) union Liveness(R)
```

In the region inferencer, this step is done in [`RegionInferenceContext::new`].

### Applying outlives constraints

Once we have computed the DAG of SCCs, we use that to structure out
entire computation. If we have an edge `S1 -> S2` between two SCCs,
that means that `Values(S1) >= Values(S2)` must hold. So, to compute
the value of `S1`, we first compute the values of each successor `S2`.
Then we simply union all of those values together. To use a
quasi-iterator-like notation:

```
Values(S1) =
  s1.successors()
    .map(|s2| Values(s2))
    .union()
```

In the code, this work starts in the [`propagate_constraints`]

> function, which iterates over all the SCCs. For each SCC S1, we

Reviewer: ```suggestion
function, which iterates over all the SCCs. For each SCC `S1`, we
```

compute its value by first computing the value of its

> successors. Since SCCs form a DAG, we don't have to be conecrned about

Reviewer: ```suggestion
successors. Since SCCs form a DAG, we don't have to be concerned about
```

cycles, though we do need to keep a set around to track whether we

> have already processed a given SCC or not. For each successor S2, once

Reviewer: ```suggestion
have already processed a given SCC or not. For each successor `S2`, once
```

> we have computed S2's value, we can union those elements into the

Reviewer: ```suggestion
we have computed `S2`'s value, we can union those elements into the
```

> value for S1. (Although we have to be careful in this process to

Reviewer: ```suggestion
value for `S1`. (Although we have to be careful in this process to
```

properly handle [higher-ranked
placeholders](./placeholders_and_universes.html). Note that the value

> for S1 already contains the liveness constraints, since they were

Reviewer: ```suggestion
for `S1` already contains the liveness constraints, since they were
```

added in [`RegionInferenceContext::new`].

> Once that process is done, we now have the "minimal value" for S1,

Reviewer: ```suggestion
Once that process is done, we now have the "minimal value" for `S1`,
```
